### PR TITLE
Add IPv6 support to `profile tcprtt`

### PR DESCRIPTION
**Pull request by eiffel-fl**
Hi.

In this contribution, I added support for IPv6 to `profile tcprtt`.

Best regards.

### CODE REVIEW

File: pkg/gadgets/profile/tcprtt/tracer/tracer_test.go [hunk 1]
```diff
@@ -505,7 +589,352 @@ func TestRunWithResult(t *testing.T) {
...
+					"expected to find histogram with local address %s", test.expectedResult.byLocalAddr)
+			} else if test.expectedResult.byRemoteAddr != "" {
+				require.True(t, rAddrFound,
+					"expected to find histogram with remote address %s", test.expectedResult.byRemoteAddr)
+			}
+		})
+	}
+}
+
+func TestRunWithResultV6(t *testing.T) {
```
Comment (blanquicet) [inline pkg/gadgets/profile/tcprtt/tracer/tracer_test.go:699]: I made this work locally using the loopback address `::1`. Did you try it? The problem with using the loopback address is that IPv6 allocates only one single loopback address, so we would need to use the same for the client and server. 

In addition, I found that some tests are reporting data for `127.0.0.1` too. I didn't investigate the reason, but here is the output:
<details>
  <summary>Tests output</summary>
  
  ```bash
$ sudo go test -run TestRunWithResultV6 ./pkg/gadgets/profile/tcprtt/...
--- FAIL: TestRunWithResultV6 (0.00s)
    --- FAIL: TestRunWithResultV6/mix_by_local_and_filter_by_laddr_v6 (4.07s)
        tracer_test.go:970: Result [1/2]:
        tracer_test.go:971: AddrType Local - Addr :: - Avg 10.731959
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 34       |**************************              |
                     8 -> 15         : 51       |****************************************|
                    16 -> 31         : 9        |*******                                 |
                    32 -> 63         : 3        |**                                      |
        tracer_test.go:970: Result [2/2]:
        tracer_test.go:971: AddrType Local - Addr 127.0.0.1 - Avg 91.954545
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 0        |                                        |
                     8 -> 15         : 0        |                                        |
                    16 -> 31         : 0        |                                        |
                    32 -> 63         : 8        |************                            |
                    64 -> 127        : 25       |****************************************|
                   128 -> 255        : 11       |*****************                       |
        tracer_test.go:977:
                Error Trace:    /inspektor-gadget/pkg/gadgets/profile/tcprtt/tracer/tracer_test.go:977
                Error:          Not equal:
                                expected: 1
                                actual  : 2
                Test:           TestRunWithResultV6/mix_by_local_and_filter_by_laddr_v6
                Messages:       invalid number of histograms
    --- FAIL: TestRunWithResultV6/mix_by_remote_and_filter_by_laddr_v6 (4.09s)
        tracer_test.go:970: Result [1/2]:
        tracer_test.go:971: AddrType Remote - Addr ::1 - Avg 72.680851
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 0        |                                        |
                     8 -> 15         : 5        |*****                                   |
                    16 -> 31         : 28       |****************************            |
                    32 -> 63         : 39       |****************************************|
                    64 -> 127        : 11       |***********                             |
                   128 -> 255        : 5        |*****                                   |
                   256 -> 511        : 5        |*****                                   |
                   512 -> 1023       : 1        |*                                       |
        tracer_test.go:970: Result [2/2]:
        tracer_test.go:971: AddrType Remote - Addr 10.0.2.2 - Avg 370.000000
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 0        |                                        |
                     8 -> 15         : 0        |                                        |
                    16 -> 31         : 0        |                                        |
                    32 -> 63         : 0        |                                        |
                    64 -> 127        : 0        |                                        |
                   128 -> 255        : 0        |                                        |
                   256 -> 511        : 2        |****************************************|
        tracer_test.go:977:
                Error Trace:    /inspektor-gadget/pkg/gadgets/profile/tcprtt/tracer/tracer_test.go:977
                Error:          Not equal:
                                expected: 1
                                actual  : 2
                Test:           TestRunWithResultV6/mix_by_remote_and_filter_by_laddr_v6
                Messages:       invalid number of histograms
    --- FAIL: TestRunWithResultV6/mix_by_remote_and_filter_by_raddr_v6 (4.09s)
        tracer_test.go:970: Result [1/2]:
        tracer_test.go:971: AddrType Remote - Addr 127.0.0.1 - Avg 68.035714
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 0        |                                        |
                     8 -> 15         : 2        |****                                    |
                    16 -> 31         : 0        |                                        |
                    32 -> 63         : 9        |*********************                   |
                    64 -> 127        : 17       |****************************************|
        tracer_test.go:970: Result [2/2]:
        tracer_test.go:971: AddrType Remote - Addr ::1 - Avg 60.263158
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 0        |                                        |
                     8 -> 15         : 6        |****                                    |
                    16 -> 31         : 6        |****                                    |
                    32 -> 63         : 56       |****************************************|
                    64 -> 127        : 24       |*****************                       |
                   128 -> 255        : 3        |**                                      |
        tracer_test.go:977:
                Error Trace:    /inspektor-gadget/pkg/gadgets/profile/tcprtt/tracer/tracer_test.go:977
                Error:          Not equal:
                                expected: 1
                                actual  : 2
                Test:           TestRunWithResultV6/mix_by_remote_and_filter_by_raddr_v6
                Messages:       invalid number of histograms
    --- FAIL: TestRunWithResultV6/by_local_addr (4.05s)
        tracer_test.go:970: Result [1/2]:
        tracer_test.go:971: AddrType Local - Addr :: - Avg 62.916667
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 0        |                                        |
                     8 -> 15         : 0        |                                        |
                    16 -> 31         : 4        |***                                     |
                    32 -> 63         : 53       |****************************************|
                    64 -> 127        : 33       |************************                |
                   128 -> 255        : 6        |****                                    |
        tracer_test.go:970: Result [2/2]:
        tracer_test.go:971: AddrType Local - Addr 127.0.0.1 - Avg 78.956522
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 0        |                                        |
                     8 -> 15         : 0        |                                        |
                    16 -> 31         : 0        |                                        |
                    32 -> 63         : 11       |************************************    |
                    64 -> 127        : 12       |****************************************|
        tracer_test.go:1035:
                Error Trace:    /inspektor-gadget/pkg/gadgets/profile/tcprtt/tracer/tracer_test.go:1035
                Error:          Should be true
                Test:           TestRunWithResultV6/by_local_addr
                Messages:       expected to find histogram with local address ::1
    --- FAIL: TestRunWithResultV6/mix_by_local_and_filter_by_raddr_v6 (4.05s)
        tracer_test.go:970: Result [1/2]:
        tracer_test.go:971: AddrType Local - Addr :: - Avg 62.916667
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 0        |                                        |
                     8 -> 15         : 0        |                                        |
                    16 -> 31         : 4        |***                                     |
                    32 -> 63         : 53       |****************************************|
                    64 -> 127        : 33       |************************                |
                   128 -> 255        : 6        |****                                    |
        tracer_test.go:970: Result [2/2]:
        tracer_test.go:971: AddrType Local - Addr 127.0.0.1 - Avg 78.956522
        tracer_test.go:972: Histogram:         µs               : count    distribution
                     0 -> 1          : 0        |                                        |
                     2 -> 3          : 0        |                                        |
                     4 -> 7          : 0        |                                        |
                     8 -> 15         : 0        |                                        |
                    16 -> 31         : 0        |                                        |
                    32 -> 63         : 11       |************************************    |
                    64 -> 127        : 12       |****************************************|
        tracer_test.go:977:
                Error Trace:    /inspektor-gadget/pkg/gadgets/profile/tcprtt/tracer/tracer_test.go:977
                Error:          Not equal:
                                expected: 1
                                actual  : 2
                Test:           TestRunWithResultV6/mix_by_local_and_filter_by_raddr_v6
                Messages:       invalid number of histograms
FAIL
FAIL    github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/profile/tcprtt/tracer  17.308s
?       github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/profile/tcprtt/types   [no test files]
FAIL
  ```
</details>

**Review by blanquicet — COMMENTED** [review]
Still some comments, and I found some issues with the IPv6 tests. However, I think it is almost ready to go. 

Would you like first to propose the bpf program to the BCC project?

File: pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c [hunk 2]
```diff
@@ -10,48 +10,108 @@
...
+		if (key.family == AF_INET6)
+			bpf_probe_read_kernel(key.addr, sizeof(BPF_CORE_READ(inet, pinet6, saddr.in6_u.u6_addr32)), BPF_CORE_READ(inet, pinet6, saddr.in6_u.u6_addr32));
+		else
+			bpf_probe_read_kernel(key.addr, sizeof(inet->inet_saddr), &inet->inet_saddr);
 	else if (targ_raddr_hist)
-		key = BPF_CORE_READ(inet, sk.__sk_common.skc_daddr);
+		if (key.family == AF_INET6)
+			bpf_probe_read_kernel(&key.addr, sizeof(sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32), &sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
+		else
+			bpf_probe_read_kernel(&key.addr, sizeof(inet->sk.__sk_common.skc_daddr), &inet->sk.__sk_common.skc_daddr);
```
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c:105]: But why `BPF_CORE_READ` is used to access one of the four cases and no the others like it is done in the filtering part?

File: pkg/gadgets/profile/tcprtt/tracer/tracer_test.go (hunk 1 above)
Comment (blanquicet) [reply pkg/gadgets/profile/tcprtt/tracer/tracer_test.go:699]: Sorry, I forgot to share the changes I did: https://github.com/inspektor-gadget/inspektor-gadget/commit/b83ee93ee8146a64f87bd2e61cb9d207c93dac71

File: pkg/gadgets/profile/tcprtt/tracer/tracer.go [hunk 3]
```diff
@@ -100,13 +102,37 @@ func (t *Tracer) parseParams(params *params.Params) error {
 		t.config.filterRemoteAddress = r
 	}
 
+	lAddrV6 := params.Get(ParamFilterLocalAddressV6).AsString()
+	if lAddrV6 != "" {
+		if lAddr != "" {
```
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/gadgets/profile/tcprtt/tracer/tracer.go:107]: But with the current code you are allowing to filter by local IPv6 and remote IPv4, which is what we don't want.

File: pkg/gadgets/profile/tcprtt/tracer/tracer.go [hunk 4]
```diff
@@ -134,14 +160,17 @@ func (t *Tracer) collectResult() ([]byte, error) {
...
 		if addressType == types.AddressTypeAll {
 			addr = types.WildcardAddress
 		} else {
-			ptr := (*[4]byte)(unsafe.Pointer(&key))
-			addr = netip.AddrFrom4(*ptr).String()
+			ipType := 4
+			if key.Family == syscall.AF_INET6 {
+				ipType = 6
+			}
+			addr = gadgets.IPStringFromBytes(key.Addr, ipType)
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/profile/tcprtt/tracer/tracer.go]: I have some doubts about how to define those types. `unsigned short` is not the same as `uint16_t` (https://stackoverflow.com/a/17693479). In many gadgets we used architecture dependent types, (int, unsigned short, etc), in others we use fixed length types. 

Specific cases:
- pid: it's uses a `pid_t typedef` defined as an `int` in the kernel. We use a combination of types in different gadgets, `pid_t`, `u32` and `__u32`. 
- `uid` and `gid`: Those are `{u,g}id_t` that is defined as `unsigned int` in the kernel. For those we mainly use `__u32` in the gadgets. 
- mount ns id: (#1215). It's unsigned in in the kernel but we use u64. 

IMO we should use exact same kernel type in the gadgets. I don't think it's very important to change now, but given you mentioned this I wanted to share my thoughts on that. 

@alban any opinions on this?

**Comment by blanquicet** [comment]
> Marking as hold until I find a good way to test it.

Did you try with the loopback address `::1`?

File: pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c [hunk 5]
```diff
@@ -10,48 +10,119 @@
...
+	default:
 		return 0;
+	}
 
 	if (targ_laddr_hist)
-		key = BPF_CORE_READ(inet, inet_saddr);
+		if (key.family == AF_INET6)
+			bpf_probe_read_kernel(key.addr, sizeof(key.addr), BPF_CORE_READ(inet, pinet6, saddr.in6_u.u6_addr8));
+		else
+			bpf_probe_read_kernel(key.addr, sizeof(inet->inet_saddr), &inet->inet_saddr);
```
Comment (blanquicet) [inline pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c:117]:
```suggestion
			bpf_probe_read_kernel(&key.addr, sizeof(key.addr), BPF_CORE_READ(inet, pinet6, saddr.in6_u.u6_addr8));
		else
			bpf_probe_read_kernel(&key.addr, sizeof(inet->inet_saddr), BPF_CORE_READ(inet, inet_saddr));
```
?

File: pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c [hunk 6]
```diff
@@ -10,48 +10,119 @@
...
 	else if (targ_raddr_hist)
-		key = BPF_CORE_READ(inet, sk.__sk_common.skc_daddr);
+		if (key.family == AF_INET6)
+			bpf_probe_read_kernel(&key.addr, sizeof(key.addr), &sk->__sk_common.skc_v6_daddr.in6_u.u6_addr8);
+		else
+			bpf_probe_read_kernel(&key.addr, sizeof(inet->sk.__sk_common.skc_daddr), &inet->sk.__sk_common.skc_daddr);
```
Comment (blanquicet) [inline pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c:122]:
```suggestion
			bpf_probe_read_kernel(&key.addr, sizeof(key.addr), BPF_CORE_READ(sk, __sk_common.skc_v6_daddr.in6_u.u6_addr8));
		else
			bpf_probe_read_kernel(&key.addr, sizeof(inet->sk.__sk_common.skc_daddr), BPF_CORE_READ(sk, __sk_common.skc_daddr));
```

?

File: pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c [hunk 7]
```diff
@@ -10,48 +10,119 @@
...
+		if (key.family == AF_INET6)
+			bpf_probe_read_kernel(&key.addr, sizeof(key.addr), BPF_CORE_READ(inet, pinet6, saddr.in6_u.u6_addr8));
+		else
+			bpf_probe_read_kernel(&key.addr, sizeof(inet->inet_saddr), &inet->inet_saddr);
 	else if (targ_raddr_hist)
-		key = BPF_CORE_READ(inet, sk.__sk_common.skc_daddr);
+		if (key.family == AF_INET6)
+			bpf_probe_read_kernel(&key.addr, sizeof(key.addr), BPF_CORE_READ(sk, __sk_common.skc_v6_daddr.in6_u.u6_addr8));
+		else
+			bpf_probe_read_kernel(&key.addr, sizeof(inet->sk.__sk_common.skc_daddr), &sk->__sk_common.skc_daddr);
```
Comment (blanquicet) [inline pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c:122]: nit: Just for consistency with the way we are accessing it:

```suggestion
			bpf_probe_read_kernel(&key.addr, sizeof(sk->__sk_common.skc_daddr), &sk->__sk_common.skc_daddr);
```

File: pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c (hunk 6 above)
Comment (blanquicet) [reply pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c]: Will it still be fine for CO-RE? (Sorry if it is a silly question)

**Review by blanquicet — APPROVED** [review]
LGTM! Thanks for working on this and managing all comments.

File: pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c (hunk 6 above)
Comment (blanquicet) [reply pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c]: @mauriciovasquezbernal @alban any thought/comment?

Comment (mauriciovasquezbernal) [reply pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c]: I think it's fine to use plain `bpf_probe_read_kernel` given that we're using vmlinux.h that has the `__attribute__((preserve_access_index))` attribute:

"But if you were to use plain old bpf_probe_read_kernel() helper directly, if the struct has preserve_access_index attribute, such probe reads become CO-RE-relocated as well. So, in short, it's always a good idea to specify this attribute." from https://nakryiko.com/posts/bpf-core-reference-guide/#the-missing-manual.

**Pull request merged**